### PR TITLE
Refactor /whois and /checkpunishment

### PR DESCRIPTION
**Pull request by bumbadadabum**
Also add a Punishment monitor

### CODE REVIEW

**Comment by Zarel** [comment]
Also I think we agreed to autolock people on 2 roombans within 10 minutes.

**Comment by panpawn** [comment]
Um, did you test this... at all, @bumbadadabum?

**Comment by panpawn** [comment]
Also, ugh, why was this merged so soon :|

File: punishments.js [hunk 1]
```diff
@@ -984,3 +990,58 @@ Punishments.isRoomBanned = function (user, roomid) {
...
+
+/**
+ * Notifies staff if a user has three or more room punishments.
+ *
+ * @param {User} user
+ */
+Punishments.monitorRoomPunishments = function (user) {
+	let punishments = Punishments.getRoomPunishments(user);
+
+	if (punishments.length >= 3) {
```
Comment (panpawn) [inline punishments.js]: ?????

``punishments`` can be undefined here....

File: rooms.js [hunk 2]
```diff
@@ -199,6 +199,9 @@ class Room {
 		this.runMuteTimer();
 
 		user.updateIdentity(this.id);
+
+		Punishments.monitorRoomPunishments(user);
```
Comment (panpawn) [inline rooms.js]: This should be a ``Config`` setting, for sure

File: punishments.js [hunk 3]
```diff
@@ -984,3 +990,58 @@ Punishments.isRoomBanned = function (user, roomid) {
...
+			if (!punishDesc) punishDesc = `punished`;
+			if (punishUserid !== user.userid) punishDesc += ` as ${punishUserid}`;
+
+			if (reason) punishDesc += `: ${reason}`;
+			return `${room} (${punishDesc})`;
+		}).join(', ');
+
+		Monitor.log(`[PunishmentMonitor] ${user.name} currently has punishments in ${punishments.length} rooms: ${punishmentText}`);
+	}
+};
```
Comment (panpawn) [inline punishments.js]: why did you delete a line here?

**Comment by panpawn** [comment]
This will count all mutes and roombans in all rooms, I'm pretty sure it should only be public rooms, let __alone__ groupchats....

**Comment by panpawn** [comment]
Okay, this definitely should of been tested more - these kinds of mistakes should of been obvious with minimal testing:

File: chat-plugins/info.js [hunk 4]
```diff
@@ -140,37 +140,24 @@ exports.commands = {
...
-				}
-				let punishDesc = ``;
-				if (punishment) {
-					const [punishType, userid, expireTime, reason] = punishment;
-					punishDesc = Punishments.roomPunishmentTypes.get(punishType);
-					if (!punishDesc) punishDesc = `punished`;
-					if (userid !== targetUser.userid) punishDesc += ` as ${userid}`;
+			let punishments = Punishments.getRoomPunishments(user);
+
+			if (punishments.length) {
```
Comment (panpawn) [inline chat-plugins/info.js]: ``punishments`` is USUALLY undefined here, so for a user whose not punished, this will crash.

**Comment by bumbadadabum** [comment]
Well, I messed up somewhere along the line and a rather crucial line of code disappeared.

**Comment by bumbadadabum** [comment]
>This will count all mutes and roombans in all rooms, I'm pretty sure it should only be public rooms, let alone groupchats....

It won't. If you read the code properly you'd know that. You'd also know the cause of the crashes being a single line. Although you were sortof right in that there were issues, your reasoning for them is far off.

**Comment by panpawn** [comment]
>It won't. 

Yea, it won't _now_: https://github.com/Zarel/Pokemon-Showdown/commit/68e49f82cb98770be208a8be16002a8532c998c0

>your reasoning for them is far off.

My reasoning for them was a lack of testing THESE changes that you proposed in THIS pull request, that's not far off at all.  

Anyways, even though this now doesn't _crash_, there are still things that should probably be done with this before it actually goes live on main, or any server.

**Comment by bumbadadabum** [comment]
>My reasoning for them was a lack of testing THESE changes that you proposed in THIS pull request, that's not far off at all.

The return probably got lost during some last-minute refactoring. You seemed to pin all the problems elsewhere, even though it seemed to me to be an obvious oversight that could have been handled in a single comment with "I think you accidentally a return statement".

>Anyways, even though this now doesn't crash, there are still things that should probably be done with this before it actually goes live on main, or any server.

I only see something about a config option in the comments section here, which is possible to add, although I didn't really see it as major.

File: chat.js [hunk 5]
```diff
@@ -938,9 +938,10 @@ Chat.toDurationString = function (number, options) {
 	const parts = [date.getUTCFullYear() - 1970, date.getUTCMonth(), date.getUTCDate() - 1, date.getUTCHours(), date.getUTCMinutes(), date.getUTCSeconds()];
 	const unitNames = ["second", "minute", "hour", "day", "month", "year"];
 	const positiveIndex = parts.findIndex(elem => elem > 0);
+	const limit = (options && options.limit ? options.limit : 0);
```
Comment (Slayer95) [inline chat.js]: This is an odd API...

There is a natural choice in the ECMA issue linked in the comment above:

```
minUnit: 'second', // millisecond | second | minute | hour
maxUnit: 'hour' // millisecond | second | minute | hour
```

Comment (Slayer95) [inline chat.js]: Hmm, then what about calling it ``precision``?

Comment (bumbadadabum) [inline chat.js]: Possible, but I like "limit" better personally. Maybe @Zarel can chime in.

Comment (panpawn) [inline chat.js]: https://github.com/Zarel/Pokemon-Showdown/pull/2971
